lib/matplotlib/figure.py

```diff
@@ -1596,25 +1596,37 @@ def subplots(self, nrows=1, ncols=1, sharex=False, sharey=False,
             return axarr
 
     def _remove_ax(self, ax):
-        def _reset_loc_form(axis):
+        def _reset_loc_from(axis):
             # Set the formatters and locators to be associated with axis
             # (where previously they may have been associated with another
             # Axis isntance)
+            #
+            # Because set_major_formatter() etc. force isDefault_* to be False,
+            # we have to manually check if the original formatter was a
+            # default and manually set isDefault_* if that was the case.
             majfmt = axis.get_major_formatter()
-            if not majfmt.axis.isDefault_majfmt:
-                axis.set_major_formatter(majfmt)
+            isDefault = majfmt.axis.isDefault_majfmt
+            axis.set_major_formatter(majfmt)
+            if isDefault:
+                majfmt.axis.isDefault_majfmt = True
 
```

Comment:
This doesn't make any sense to me.  You get the formatter for `axis` and then you set the formatter for `axis` to that formatter, and then you set the isDefault to the value axis had before.  I don't see how this code is doing anything.

Reply:
I think the confusion here stems from the mappings of formatters to axis.
- Axis > Formatter is a one to one mapping
- Formatter > Axis is a one to many mapping

So a single `Formatter` is linked to a single `Axis`, but many other `Axis` can make use of that `Formatter`. I agree this is a nightmare to understand, but without re-doing the internals of how this works it is how it is.

So this code is getting the formatter that `axis` is using, (that it might be borrowing) and then explicitly associating that formatter with `axis` (so it is no longer borrowing it).